BoardGameTrainer/Application/MainWindow.cs

```diff
@@ -14,151 +13,187 @@ enum WindowState
             mainVBox.PackStart(titleAndContentVBox, true, true, 0);
-            titleAndContentVBox.Show();
-            this.Add(mainVBox);
             mainVBox.Show();
-            newGameButton.Show();
-            restartButton.Show();          
+            Add(mainVBox);
         }
 
         private void BoardImageClickHandler(object sender, ButtonPressEventArgs args)
         {
-            if(windowState != WindowState.ProcessMovement)
+            if (WindowState != WindowState.ProcessMovement)
             {
-                windowState = WindowState.ProcessMovement;
-                int minDimention = Math.Min(boardImage.AllocatedWidth, boardImage.AllocatedHeight);
-                int xOffset = (boardImage.AllocatedWidth - minDimention) / 2;
-                int yOffset = (boardImage.AllocatedHeight - minDimention) / 2;
-                x = (args.Event.X - xOffset) / minDimention;
-                y = (args.Event.Y - yOffset) / minDimention;
-                Console.WriteLine($"Button Pressed at {x}, {y}");
-
-                eventsQueue.Add(PerformMovement);
+                WindowState = WindowState.ProcessMovement;
+                int minDimention = Math.Min(BoardImage.AllocatedWidth, BoardImage.AllocatedHeight);
+                int xOffset = (BoardImage.AllocatedWidth - minDimention) / 2;
+                int yOffset = (BoardImage.AllocatedHeight - minDimention) / 2;
+                X = (args.Event.X - xOffset) / minDimention;
+                Y = (args.Event.Y - yOffset) / minDimention;
+
+                PerformMovement();
+            }
+        }
+
+        internal void StartGame()
+        {
+            ResetToken();
+            WindowState = WindowState.Idle;
+            if (GameManager!.HumanPlayers[Player.One])
+            {
+                WindowState = WindowState.ComputeHints;
+                EventsQueue.Add(ComputeHints);
+            }
+            else
+            {
+                WindowState = WindowState.ProcessMovement;
+                EventsQueue.Add(PerformAiMovement);
             }
         }
 
         private void PerformMovement()
         {
-            CancellationToken token = ResetToken();
-            (GameResult gameResult, bool actionPerformed) = application.gameManager.HandleMovement(x, y, application.isPlayer2Ai);
-            if (gameResult != GameResult.InProgress)
-                application.gameManager = new DefaultGameManager(gameResult);
-            Gtk.Application.Invoke(delegate {
-                boardImage.QueueDraw();
-                
-            });
-            if (actionPerformed)
+            (GameResult gameResult, bool isActionPerformed) = GameManager!.HandleMovement(X, Y);
+            Application.Invoke(delegate { BoardImage.QueueDraw(); });
+            if (isActionPerformed && gameResult == GameResult.InProgress)
             {
-                if (application.isPlayer2Ai)
+                CancellationToken token = ResetToken();
```

Comment:
czy to, że resetujemy token dopiero tutaj nie oznacza, że mogą się nam wyświetlić jakieś przedawnione podpowiedzi zanim się wykona ten ruch?
Z drugiej strony poprzednie rozwiązanie chyba nadgorliwie cancellowało liczenie podpowiedzi